aimmo-game/simulation/custom_map.py

```diff
@@ -0,0 +1,183 @@
+import abc
+import math
+
+from simulation.levels.levels import LEVELS
+from simulation.levels.completion_checks import COMPLETION_CHECKS
+from simulation.levels.decoders import *
+
+from pprint import pprint
+
+from simulation.location import Location
+from simulation.game_state import GameState
+from simulation.world_map import WorldMap
+
+from simulation.world_map import WorldMapStaticSpawnDecorator
+from simulation.world_map import DEFAULT_LEVEL_SETTINGS
+from simulation.world_map import Cell
+
+import sys
+current_module = sys.modules[__name__]
+
+class BaseGenerator(object):
+    """
+        A map generator that exposes a game state and a check for level completion.
+
+        API:
+            - contructor(setting)
+                - a set of basic settings that the map uses at generation
+                - see DEFAULT_LEVEL_SETTINGS in simulation.world_map
+            - get_game_state(avatar_manager)
+                - exposes a game state used by the turn manager daemon
+                - for details see GameState
+            - check_complete(game_state)
+                - function to check if a map is "complete"
+                - the turn manager runs the action for each avatar, then runs check_complete afterwards
+            @abstract
+            - get_map
+                - returns the generated map
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, settings):
+        self.settings = settings
+
+    def get_game_state(self, avatar_manager):
+        return GameState(self.get_map(), avatar_manager, self.check_complete)
+
+    def check_complete(self, game_state):
+        return False
```

Comment:
huh?

Reply:
The default behavior for the check complete is to to never end the level. We override that in levels with the specific checks.

Reply:
try abstracting it?

Reply:
@CelineBoudier  @danalex97  I'm not sure about python as it's still quite new to me, but I found the readibility of abstract methods/classes really difficult. Maybe we should use something like '@abc.abstractmethod'?

https://pymotw.com/3/abc/

Reply:
Yeah, that's how I generally do it. But I wanted to make that a method with a default behavior, as all the classes EmptyMapGenerator, Main and JSONLevelGenerator use check_complete. I think is better this way.

Why we have the parameter game state is that we might want some custom completion checks that do something like check the score and so on.